astropy/table/bst.py

```diff
@@ -0,0 +1,614 @@
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+from __future__ import (absolute_import, division, print_function,
+                        unicode_literals)
+import operator
+import numpy as np
+from ..extern.six.moves import zip
+
+
+class MaxValue(object):
+    '''
+    Represents an infinite value for purposes
+    of tuple comparison.
+    '''
+    def __gt__(self, other):
+        return True
+
+    def __ge__(self, other):
+        return True
+
+    def __lt__(self, other):
+        return False
+
+    def __le__(self, other):
+        return False
+
+    def __repr__(self):
+        return "MAX"
+
+    __le__ = __lt__
+    __ge__ = __gt__
+    __str__ = __repr__
+
+
+class MinValue(object):
+    '''
+    The opposite of MaxValue, i.e. a representation of
+    negative infinity.
+    '''
+    def __lt__(self, other):
+        return True
+
+    def __le__(self, other):
+        return True
+
+    def __gt__(self, other):
+        return False
+
+    def __ge__(self, other):
+        return False
+
+    def __repr__(self):
+        return "MIN"
+
+    __le__ = __lt__
+    __ge__ = __gt__
+    __str__ = __repr__
+
+
+class Epsilon(object):
+    '''
+    Represents the "next largest" version of a given value,
+    so that for all valid comparisons we have
+    x < y < Epsilon(y) < z whenever x < y < z and x, z are
+    not Epsilon objects.
+
+    Parameters
+    ----------
+    val : object
+        Original value
+    '''
+
+    def __init__(self, val):
+        self.val = val
+
+    def __lt__(self, other):
+        if self.val == other:
+            return False
+        return self.val < other
+
+    def __gt__(self, other):
+        if self.val == other:
+            return True
+        return self.val > other
+
+    def __eq__(self, other):
+        return False
+
+    def __repr__(self):
+        return repr(self.val) + " + epsilon"
+
+class Node(object):
+    '''
+    An element in a binary search tree, containing
+    a key, data, and references to children nodes and
+    a parent node.
+
+    Parameters
+    ----------
+    key : tuple
+        Node key
+    data : list or int
+        Node data
+    '''
+    __lt__ = lambda x, y: x.key < y.key
+    __le__ = lambda x, y: x.key <= y.key
+    __eq__ = lambda x, y: x.key == y.key
+    __ge__ = lambda x, y: x.key >= y.key
+    __gt__ = lambda x, y: x.key > y.key
+    __ne__ = lambda x, y: x.key != y.key
+
+    # each node has a key and data list
+    def __init__(self, key, data):
+        self.key = key
+        self.data = data if isinstance(data, list) else [data]
+        self.left = None
+        self.right = None
+        self.parent = None
```

Comment:
On second thought, is the `self.parent` reference necessary?  I don't see how it's being used anywhere.  I also would worry about how the resulting reference cycles will affect garbage collection performance.  Better to avoid if it's not needed.